src/agent/src/image.rs

```diff
@@ -0,0 +1,382 @@
+// Copyright (c) 2021 Alibaba Cloud
+// Copyright (c) 2021, 2023 IBM Corporation
+// Copyright (c) 2022 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::collections::HashMap;
+use std::env;
+use std::fs;
+use std::path::Path;
+use std::sync::Arc;
+
+use anyhow::{anyhow, Context, Result};
+use image_rs::image::ImageClient;
+use tokio::sync::Mutex;
+
+use crate::rpc::CONTAINER_BASE;
+use crate::AGENT_CONFIG;
+
+// A marker to merge container spec for images pulled inside guest.
+const ANNO_K8S_IMAGE_NAME: &str = "io.kubernetes.cri.image-name";
+const KATA_IMAGE_WORK_DIR: &str = "/run/image/";
+const CONFIG_JSON: &str = "config.json";
+const KATA_PAUSE_BUNDLE: &str = "/pause_bundle";
+
+const K8S_CONTAINER_TYPE_KEYS: [&str; 2] = [
+    "io.kubernetes.cri.container-type",
+    "io.kubernetes.cri-o.ContainerType",
+];
+
+#[rustfmt::skip]
+lazy_static! {
+    pub static ref IMAGE_SERVICE: Mutex<Option<ImageService>> = Mutex::new(None);
+}
+
+// Convenience function to obtain the scope logger.
+fn sl() -> slog::Logger {
+    slog_scope::logger().new(o!("subsystem" => "cgroups"))
+}
+
+#[derive(Clone)]
+pub struct ImageService {
+    image_client: Arc<Mutex<ImageClient>>,
+    images: Arc<Mutex<HashMap<String, String>>>,
+}
+
+impl ImageService {
+    pub fn new() -> Self {
+        env::set_var("KATA_IMAGE_WORK_DIR", KATA_IMAGE_WORK_DIR);
+
+        Self {
+            image_client: Arc::new(Mutex::new(ImageClient::default())),
+            images: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+
+    /// Get the singleton instance of image service.
+    pub async fn singleton() -> Result<ImageService> {
+        IMAGE_SERVICE
+            .lock()
+            .await
+            .clone()
+            .ok_or_else(|| anyhow!("image service is uninitialized"))
+    }
```

Comment:
I guess `singleton()` is intended to give a reference/pointer to the static variable `IMAGE_SERVICE`.
    
However, the current `clone()` after `lock()` will copy the whole `Option<ImageService>` in memory every time `singleton()` is called. This means that if tens of image pulling requests are sent to kata-agent, tens of `Option<ImageService>` will be allocated inside the memory which will occupy extra memory.

I suggest to add two `pub` functions in `image.rs`:
1. `merge_bundle_oci()` to be called https://github.com/kata-containers/kata-containers/pull/8484/files#diff-f87c1c212bba3abd0b07bafe69d5b01ad00dfe2c01bed585e2a75fcbb6442b89R224-R228
2. `init_image_service()` to be called https://github.com/kata-containers/kata-containers/pull/8484/files#diff-f87c1c212bba3abd0b07bafe69d5b01ad00dfe2c01bed585e2a75fcbb6442b89R1616-R1620

Then the operations upon the static variable `IMAGE_SERVICE` could be implemented inside these two functions without expose any inner details to outer module like `rpc.rs`. This might help to decouple.

Reply:
Thanks for your advice. I thinks we can make these changes in the following PR.